<!DOCTYPE html>



  


<html class="theme-next gemini use-motion" lang="en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.4" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.4">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.4" color="#222">





  <meta name="keywords" content="edgeR,diﬀerential gene analysis," />










<meta name="description" content="1. 简介edgeR作用对象是count文件，rows 代表基因，行代表文库，count代表的是比对到每个基因的reads数目。它主要关注的是差异表达分析，而不是定量基因表达水平。 edgeR works on a table of integer read counts, with rows corresponding to genes and columns to independent li">
<meta property="og:type" content="article">
<meta property="og:title" content="5-RNAseq--无重复差异基因分析（edgeR包的使用）">
<meta property="og:url" content="http://yoursite.com/2020/03/19/5-RNAseq--%E6%97%A0%E9%87%8D%E5%A4%8D%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%88%86%E6%9E%90%EF%BC%88edgeR%E5%8C%85%E7%9A%84%E4%BD%BF%E7%94%A8%EF%BC%89/index.html">
<meta property="og:site_name" content="小钱的个人博客">
<meta property="og:description" content="1. 简介edgeR作用对象是count文件，rows 代表基因，行代表文库，count代表的是比对到每个基因的reads数目。它主要关注的是差异表达分析，而不是定量基因表达水平。 edgeR works on a table of integer read counts, with rows corresponding to genes and columns to independent li">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/3194654-fa83a5ee0fe58872.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/3194654-d1f8202b26ca0a6b.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/3194654-8b5ece418efe7862.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/3194654-d48f8e81d2527524.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/3194654-7c9c6b8d2527a3a0.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/3194654-15d86de730b879bc.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/3194654-1b078402234a1257.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">
<meta property="article:published_time" content="2020-03-19T14:12:53.000Z">
<meta property="article:modified_time" content="2020-03-31T06:32:09.090Z">
<meta property="article:author" content="WENJU QIAN">
<meta property="article:tag" content="edgeR">
<meta property="article:tag" content="diﬀerential gene analysis">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://upload-images.jianshu.io/upload_images/3194654-fa83a5ee0fe58872.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Gemini',
    version: '5.1.4',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":true,"scrollpercent":true,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://yoursite.com/2020/03/19/5-RNAseq--无重复差异基因分析（edgeR包的使用）/"/>





  <title>5-RNAseq--无重复差异基因分析（edgeR包的使用） | 小钱的个人博客</title>
  








<meta name="generator" content="Hexo 4.2.0"></head>


<script type="text/javascript" src="//cdn.bootcss.com/canvas-nest.js/1.0.0/canvas-nest.min.js"></script>


<script src="/live2d-widget/autoload.js"></script>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta custom-logo">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">小钱的个人博客</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">WENJU QIAN's personal website</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            About
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            Tags
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            Categories
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            Archives
          </a>
        </li>
      
        
        <li class="menu-item menu-item-love">
          <a href="/love/s/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-heartbeat"></i> <br />
            
            love
          </a>
        </li>
      
        
        <li class="menu-item menu-item-sitemap">
          <a href="/sitemap.xml" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-sitemap"></i> <br />
            
            Sitemap
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />
            
            Search
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2020/03/19/5-RNAseq--%E6%97%A0%E9%87%8D%E5%A4%8D%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%88%86%E6%9E%90%EF%BC%88edgeR%E5%8C%85%E7%9A%84%E4%BD%BF%E7%94%A8%EF%BC%89/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="WENJU QIAN">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/head_logo.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="小钱的个人博客">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">5-RNAseq--无重复差异基因分析（edgeR包的使用）</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2020-03-19T22:12:53+08:00">
                2020-03-19
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/RNA-seq/" itemprop="url" rel="index">
                    <span itemprop="name">RNA-seq</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/2020/03/19/5-RNAseq--%E6%97%A0%E9%87%8D%E5%A4%8D%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%88%86%E6%9E%90%EF%BC%88edgeR%E5%8C%85%E7%9A%84%E4%BD%BF%E7%94%A8%EF%BC%89/#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2020/03/19/5-RNAseq--%E6%97%A0%E9%87%8D%E5%A4%8D%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%88%86%E6%9E%90%EF%BC%88edgeR%E5%8C%85%E7%9A%84%E4%BD%BF%E7%94%A8%EF%BC%89/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          

          
            <span class="post-meta-divider">|</span>
            <span class="page-pv"><i class="fa fa-file-o"></i>
            <span class="busuanzi-value" id="busuanzi_value_page_pv" ></span>次
            </span>
          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article&#58;</span>
                
                <span title="Words count in article">
                  2k
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h1 id="1-简介"><a href="#1-简介" class="headerlink" title="1. 简介"></a>1. 简介</h1><p>edgeR作用对象是count文件，rows 代表基因，行代表文库，count代表的是比对到每个基因的reads数目。它主要关注的是差异表达分析，而不是定量基因表达水平。</p>
<p>edgeR works on a table of integer read counts, with rows corresponding to genes and columns to independent libraries. The counts represent the total number of reads aligning to each gene (or other genomic locus).edgeR is concerned with diﬀerential expression analysis rather than with the quantiﬁcation of expression levels. It is concerned with relative changes in expression levels between conditions,but not directly with estimating absolute expression levels.</p>
<p>edgeR作用的是真实的比对统计，因此不建议用预测的转录本</p>
<p>Note that edgeR is designed to work with actual read counts. We not recommend that predicted transcript abundances are input the edgeR in place of actual counts.</p>
<p>归一化原因：</p>
<blockquote>
<p>技术原因影响差异表达分析：<br>1）Sequencing depth：统计测序深度(即代表的是library size)；<br>2）RNA composition：个别异常高表达基因导致其它基因采样不足<br>3）GC content： sample-speciﬁc eﬀects for GC-content can be detected<br>4）sample-speciﬁc eﬀects for gene length have been detected</p>
</blockquote>
<p><strong>注意：edgeR必须是原始表达量，而不能是rpkm等矫正过的。</strong><br>Note that normalization in edgeR is model-based, and the original read counts are not themselves transformed. This means that users should not transform the read counts in any way before inputing them to edgeR. For example, users should not enter RPKM or FPKM values to edgeR in place of read counts. Such quantities will prevent edgeR from correctly estimating the mean-variance relationship in the data, which is a crucial to the statistical strategies underlying edgeR.Similarly, users should not add artiﬁcial values to the counts before inputing them to edgeR.<br>个人是不太推荐没有重复的差异表达分析，因为毕竟统计学上的p值是为了证明两个样本的差异是真实存在而不是抽样误差导致的，<br>因此每当别人提问的时候, 我个人的建议就是定性看看倍数变化吧. 但是如果真的强行要算p值, 其实也不是不行, edgeR就是一种选择.</p>
<h1 id="2-首先安装edgeR-包"><a href="#2-首先安装edgeR-包" class="headerlink" title="2. 首先安装edgeR 包"></a>2. 首先安装edgeR 包</h1><figure class="highlight cmake"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#如果没有安装BiocMaRnager则先安装BiocManager，之后通过BiocManager安装edgeR包</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">"BiocManager"</span>, quietly = <span class="keyword">TRUE</span>))</span><br><span class="line">    <span class="keyword">install</span>.packages(<span class="string">"BiocManager"</span>)</span><br><span class="line"></span><br><span class="line">BiocManager::<span class="keyword">install</span>(<span class="string">"edgeR"</span>)</span><br></pre></td></tr></table></figure>
<p>安装结束之后开始处理文件</p>
<h1 id="3-矩阵构建及差异分析"><a href="#3-矩阵构建及差异分析" class="headerlink" title="3. 矩阵构建及差异分析"></a>3. 矩阵构建及差异分析</h1><p>需要构建2个矩阵：1、表达矩阵；2、分组矩阵( 实验设计)；</p>
<h2 id="3-1-表达矩阵"><a href="#3-1-表达矩阵" class="headerlink" title="3.1 表达矩阵"></a>3.1 表达矩阵</h2><h3 id="3-1-1-读取文件"><a href="#3-1-1-读取文件" class="headerlink" title="3.1.1 读取文件"></a>3.1.1 读取文件</h3><figure class="highlight angelscript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"># 首先读取counts文件之后查看count文件前<span class="number">6</span>行</span><br><span class="line">&gt; rawdata &lt;- read.csv(file = <span class="string">"C://Users/My/Desktop/diff_name"</span>,header = T,<span class="built_in">string</span>sAsFactors = F)</span><br><span class="line"># 查看读取的diff_name文件</span><br><span class="line">&gt; head(rawdata)</span><br><span class="line">  X    ensembl_gene_id               gene_id control_W55 test_K54 external_gene_name</span><br><span class="line"><span class="number">1</span> <span class="number">1</span> ENSMUSG00000000001  ENSMUSG00000000001<span class="number">.4</span>           <span class="number">7</span>       <span class="number">11</span>              Gnai3</span><br><span class="line"><span class="number">2</span> <span class="number">2</span> ENSMUSG00000000003 ENSMUSG00000000003<span class="number">.15</span>           <span class="number">0</span>        <span class="number">0</span>               Pbsn</span><br><span class="line"><span class="number">3</span> <span class="number">3</span> ENSMUSG00000000028 ENSMUSG00000000028<span class="number">.15</span>           <span class="number">1</span>        <span class="number">0</span>              Cdc45</span><br><span class="line"><span class="number">4</span> <span class="number">4</span> ENSMUSG00000000031 ENSMUSG00000000031<span class="number">.16</span>           <span class="number">0</span>        <span class="number">2</span>                H19</span><br><span class="line"><span class="number">5</span> <span class="number">5</span> ENSMUSG00000000037 ENSMUSG00000000037<span class="number">.17</span>           <span class="number">0</span>        <span class="number">0</span>              Scml2</span><br><span class="line"><span class="number">6</span> <span class="number">6</span> ENSMUSG00000000049 ENSMUSG00000000049<span class="number">.11</span>          <span class="number">54</span>       <span class="number">33</span>               Apoh</span><br></pre></td></tr></table></figure>
<p>读取完数据之后我们先预处理一下数据，比如我只想要ensembl_gene_id、control_w55、tese_k54、external_gene_name这几列，并调整一下顺序。</p>
<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">&gt; swap_rawdata &lt;- cbind(rawdata$ensembl_gene_id,rawdata$external_gene_name,rawdata$control_W55,rawdata$test_K54)</span><br><span class="line">&gt; head(swap_rawdata)</span><br><span class="line">     [,<span class="number">1</span>]                 [,<span class="number">2</span>]    [,<span class="number">3</span>] [,<span class="number">4</span>]</span><br><span class="line">[<span class="number">1</span>,] <span class="string">"ENSMUSG00000000001"</span> <span class="string">"Gnai3"</span> <span class="string">"7"</span>  <span class="string">"11"</span></span><br><span class="line">[<span class="number">2</span>,] <span class="string">"ENSMUSG00000000003"</span> <span class="string">"Pbsn"</span>  <span class="string">"0"</span>  <span class="string">"0"</span> </span><br><span class="line">[<span class="number">3</span>,] <span class="string">"ENSMUSG00000000028"</span> <span class="string">"Cdc45"</span> <span class="string">"1"</span>  <span class="string">"0"</span> </span><br><span class="line">[<span class="number">4</span>,] <span class="string">"ENSMUSG00000000031"</span> <span class="string">"H19"</span>   <span class="string">"0"</span>  <span class="string">"2"</span> </span><br><span class="line">[<span class="number">5</span>,] <span class="string">"ENSMUSG00000000037"</span> <span class="string">"Scml2"</span> <span class="string">"0"</span>  <span class="string">"0"</span> </span><br><span class="line">[<span class="number">6</span>,] <span class="string">"ENSMUSG00000000049"</span> <span class="string">"Apoh"</span>  <span class="string">"54"</span> <span class="string">"33"</span></span><br><span class="line"># 得到的这个swap_rawdata是一个matrix，如果想要让其变为data frame</span><br><span class="line">&gt; swap_rawdata &lt;- data.frame(swap_rawdata)</span><br><span class="line"># 查看一下是否转化成功</span><br><span class="line">&gt; head(swap_rawdata)</span><br><span class="line">     ensembl_gene_id external_gene_name control_W55 test_K54</span><br><span class="line"><span class="number">1</span> ENSMUSG00000000001              Gnai3           <span class="number">7</span>       <span class="number">11</span></span><br><span class="line"><span class="number">2</span> ENSMUSG00000000003               Pbsn           <span class="number">0</span>        <span class="number">0</span></span><br><span class="line"><span class="number">3</span> ENSMUSG00000000028              Cdc45           <span class="number">1</span>        <span class="number">0</span></span><br><span class="line"><span class="number">4</span> ENSMUSG00000000031                H19           <span class="number">0</span>        <span class="number">2</span></span><br><span class="line"><span class="number">5</span> ENSMUSG00000000037              Scml2           <span class="number">0</span>        <span class="number">0</span></span><br><span class="line"><span class="number">6</span> ENSMUSG00000000049               Apoh          <span class="number">54</span>       <span class="number">33</span></span><br><span class="line"># 转化完转化先存一份csv文件在电脑里，便于之后用电脑查看</span><br><span class="line">&gt; write.csv(x = swap_rawdata,file = <span class="string">"C://Users/My/Desktop/swap_rawdata.csv"</span>)</span><br><span class="line"># 存完之后直接从电脑导入你刚存的文件，这样做可以避免出现numeric数据框变成factor形式</span><br><span class="line">&gt; swap_rawdata &lt;- read.table(<span class="string">"swap_rawdata.csv"</span>,header = T,sep = <span class="string">","</span>)</span><br><span class="line"># 查看</span><br><span class="line">&gt; head(swap_rawdata)</span><br><span class="line">  X    ensembl_gene_id external_gene_name control_W55 test_K54</span><br><span class="line"><span class="number">1</span> <span class="number">1</span> ENSMUSG00000000001              Gnai3           <span class="number">7</span>       <span class="number">11</span></span><br><span class="line"><span class="number">2</span> <span class="number">2</span> ENSMUSG00000000003               Pbsn           <span class="number">0</span>        <span class="number">0</span></span><br><span class="line"><span class="number">3</span> <span class="number">3</span> ENSMUSG00000000028              Cdc45           <span class="number">1</span>        <span class="number">0</span></span><br><span class="line"><span class="number">4</span> <span class="number">4</span> ENSMUSG00000000031                H19           <span class="number">0</span>        <span class="number">2</span></span><br><span class="line"><span class="number">5</span> <span class="number">5</span> ENSMUSG00000000037              Scml2           <span class="number">0</span>        <span class="number">0</span></span><br><span class="line"><span class="number">6</span> <span class="number">6</span> ENSMUSG00000000049               Apoh          <span class="number">54</span>       <span class="number">33</span></span><br><span class="line">&gt; data.class(swap_rawdata[<span class="number">1</span>,<span class="number">1</span>])</span><br></pre></td></tr></table></figure>


<h3 id="3-1-2-接着构建DGEList对象"><a href="#3-1-2-接着构建DGEList对象" class="headerlink" title="3.1.2 接着构建DGEList对象"></a>3.1.2 接着构建DGEList对象</h3><p>这里因为已经有rawdata的count文件，因此直接用DGEList()函数就行了，否则要用readDGE()函数</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 首先载入edgeR 包</span></span><br><span class="line">&gt; library(edgeR)</span><br><span class="line"><span class="comment"># 构建DGEList</span></span><br><span class="line">&gt;<span class="built_in"> group </span>&lt;- 1:2</span><br><span class="line">&gt; y &lt;- DGEList(counts = swap_rawdata[,4:5],genes = swap_rawdata[,2:3],group = group)</span><br><span class="line"><span class="comment"># 查看构建完y的信息</span></span><br><span class="line">&gt; y</span><br></pre></td></tr></table></figure>
<p>查看构建DGElist的运行结果:</p>
<p> DGEList对象主要有三部分：</p>
<p>1、counts矩阵：包含的是整数counts;</p>
<p>2、samples数据框：包含的是文库(sample)信息。包含 lib.size列 ：for the library size (sequencing depth) for each sample,如果不自定义，  the library sizes will be computed from the column sums of the counts。其中还有一个group列，用于指定每个sample组信息</p>
<p>3、一个可选的数据框genes：gene的注释信息</p>
<p><img src="https://upload-images.jianshu.io/upload_images/3194654-fa83a5ee0fe58872.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt=""></p>
<p><strong>第二步： 过滤 low counts数据。</strong>与DESeq2的预过滤不同，DESeq2的预过滤只是为了改善后续运算性能，在运行过程中依旧会自动处理low count数据，edgeR需要在分析前就要排除那些low count数据，而且非常严格。从生物学角度，有生物学意义的基因的表达量必须高于某一个阈值。从统计学角度上， low count的数据不太可能有显著性差异，而且在多重试验矫正阶段还会拖后腿。 </p>
<h1 id="4-数据过滤"><a href="#4-数据过滤" class="headerlink" title="4. 数据过滤"></a>4. 数据过滤</h1><p>由于原来的表达量矩阵基因数太大, 可能存在某些基因根本没有表达, 因此需要预先过滤</p>
<figure class="highlight gml"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&gt; keep &lt;- rowSums(cpm(<span class="symbol">y</span>)&gt;<span class="number">1</span>) &gt;= <span class="number">1</span></span><br><span class="line">&gt; <span class="symbol">y</span> &lt;- <span class="symbol">y</span>[keep, , keep.lib.sizes=FALSE]</span><br><span class="line">&gt; <span class="symbol">y</span></span><br></pre></td></tr></table></figure>
<p>这部分代码的意思指的是保留在至少在一个样本里有表达的基因(CPM &gt; 1)。 基因数就从原来的55318变为15868</p>
<p><img src="https://upload-images.jianshu.io/upload_images/3194654-d1f8202b26ca0a6b.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt=""></p>
<h1 id="5-标准化"><a href="#5-标准化" class="headerlink" title="5. 标准化"></a>5. 标准化</h1><p>考虑到测序深度不同, 我们需要对其进行标准化, 避免文库大小不同导致的分析误差.</p>
<p>edgeR里默认采用TMM(trimmed mean of M-values) 对配对样本进行标准化，用到的函数是calcNormFactors</p>
<figure class="highlight gml"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">&gt; <span class="symbol">y</span> &lt;- calcNormFactors(<span class="symbol">y</span>)</span><br></pre></td></tr></table></figure>
<p>执行结果：</p>
<p><img src="https://upload-images.jianshu.io/upload_images/3194654-8b5ece418efe7862.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt=""></p>
<h1 id="6-差异表达分析"><a href="#6-差异表达分析" class="headerlink" title="6. 差异表达分析"></a>6. 差异表达分析</h1><p>不同差异表达分析工具的目标就是预测出dispersion(离散值), 有了离散值就能够计算p值. 那么dispersion怎么计算呢? edgeR给了几个方法</p>
<p>根据经验给定一个值(BCV, square-root-dispersion). edgeR给的建议是, 如果你是人类数据, 且实验做的很好(无过多的其他因素影响), 设置为0.4, 如果是遗传上相似的模式物种（这里为小鼠）, 设置为0.1 （查询edgeR的bioconductor包所得）<br>Simply pick a reasonable dispersion value, based on your experience with similar data,and use that for exactTest or glmFit. Typical values for the common BCV (square-root dispersion) for datasets arising from well-controlled experiments are 0.4 for human data,0.1 for data on genetically identical model organisms or 0.01 for technical replicates.</p>
<figure class="highlight reasonml"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt; y_bcv &lt;- y</span><br><span class="line"># 因为本次的数据使用的是小鼠的数据，所以使用<span class="number">0.1</span></span><br><span class="line">&gt; bcv &lt;- <span class="number">0.1</span></span><br><span class="line">&gt; et &lt;- exact<span class="constructor">Test(<span class="params">y_bcv</span>, <span class="params">dispersion</span> = <span class="params">bcv</span> ^ 2)</span></span><br><span class="line">&gt; gene1 &lt;- decide<span class="constructor">TestsDGE(<span class="params">et</span>, <span class="params">p</span>.<span class="params">value</span> = 0.05, <span class="params">lfc</span> = 0)</span></span><br><span class="line">&gt; head(gene1)</span><br><span class="line">&gt; summary(gene1)</span><br></pre></td></tr></table></figure>
<p>执行结果：</p>
<p><img src="https://upload-images.jianshu.io/upload_images/3194654-d48f8e81d2527524.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt=""></p>
<p>由统计结果可知，下调的基因为816个，上调的基因为572个</p>
<p>如果觉得觉得基因较多的话，可以上调bcv的值</p>
<figure class="highlight reasonml"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; y_bcv &lt;- y</span><br><span class="line">&gt; bcv &lt;- <span class="number">0.2</span></span><br><span class="line">&gt; et2 &lt;- exact<span class="constructor">Test(<span class="params">y_bcv</span>, <span class="params">dispersion</span> = <span class="params">bcv</span> ^ 2)</span></span><br><span class="line">&gt; gene2 &lt;- decide<span class="constructor">TestsDGE(<span class="params">et2</span>, <span class="params">p</span>.<span class="params">value</span> = 0.05, <span class="params">lfc</span> = 0)</span></span><br><span class="line">&gt; summary(gene2)</span><br></pre></td></tr></table></figure>
<p>执行结果：</p>
<p><img src="https://upload-images.jianshu.io/upload_images/3194654-7c9c6b8d2527a3a0.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt=""></p>
<p>由统计结果可知，下调的基因为377个，上调的基因为221个<br>与之前的结果相比，的确已经减少很多</p>
<p>将结果整理成excel表</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> 改一下gene1的名称</span></span><br><span class="line"><span class="meta">&gt;</span><span class="bash"> colnames(gene1) &lt;- <span class="string">"Signifi"</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 组合将所需要的数据组成一个新的data.frame</span></span><br><span class="line"><span class="meta">&gt;</span><span class="bash"> results &lt;- cbind(y<span class="variable">$genes</span>,y<span class="variable">$counts</span>,et<span class="variable">$table</span>,gene1)</span></span><br><span class="line"><span class="meta">&gt;</span><span class="bash"> head(results)</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 将新生成的results数据框写成一个excel数据表</span></span><br><span class="line"><span class="meta">&gt;</span><span class="bash"> write.csv(x = results,file = <span class="string">"C://Users/My/Desktop/DEresult.csv"</span>,row.names = F)</span></span><br></pre></td></tr></table></figure>
<p>执行结果：</p>
<p><img src="https://upload-images.jianshu.io/upload_images/3194654-15d86de730b879bc.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="组合成的新的data.frame表"></p>
<p>生成的excel表可以将down expressed 和 up expressed基因分开</p>
<p><img src="https://upload-images.jianshu.io/upload_images/3194654-1b078402234a1257.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="生成的excel"></p>
<p>参考文献：<a href="https://blog.csdn.net/u012110870/article/details/102804557" target="_blank" rel="noopener">https://blog.csdn.net/u012110870/article/details/102804557</a><br><a href="https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf" target="_blank" rel="noopener">https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf</a></p>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/edgeR/" rel="tag"># edgeR</a>
          
            <a href="/tags/di%EF%AC%80erential-gene-analysis/" rel="tag"># diﬀerential gene analysis</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2020/03/19/4-RNAseq--%E5%AF%B9RNAseq%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%E7%9A%84%E8%B4%A8%E9%87%8F%E6%8E%A7%E5%88%B6%EF%BC%88fastqc%EF%BC%89/" rel="next" title="4-RNAseq--reads count(htseq-count)与基因注释（bioMart）">
                <i class="fa fa-chevron-left"></i> 4-RNAseq--reads count(htseq-count)与基因注释（bioMart）
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2020/03/27/11/" rel="prev" title="临时博客">
                临时博客 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image"
                src="/images/head_logo.jpg"
                alt="WENJU QIAN" />
            
              <p class="site-author-name" itemprop="name">WENJU QIAN</p>
              <p class="site-description motion-element" itemprop="description">UCAS小硕一枚,21世纪生物搬砖工</p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/%7C%7C%20archive">
              
                  <span class="site-state-item-count">8</span>
                  <span class="site-state-item-name">posts</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">1</span>
                  <span class="site-state-item-name">categories</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">13</span>
                  <span class="site-state-item-name">tags</span>
                </a>
              </div>
            

          </nav>

          

          
            <div class="links-of-author motion-element">
                
                  <span class="links-of-author-item">
                    <a href="https://github.com/qianwenju" target="_blank" title="GitHub">
                      
                        <i class="fa fa-fw fa-github"></i>GitHub</a>
                  </span>
                
                  <span class="links-of-author-item">
                    <a href="mailto:qianwenju2018@sibs.ac.cn" target="_blank" title="E-Mail">
                      
                        <i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  </span>
                
            </div>
          

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#1-简介"><span class="nav-number">1.</span> <span class="nav-text">1. 简介</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#2-首先安装edgeR-包"><span class="nav-number">2.</span> <span class="nav-text">2. 首先安装edgeR 包</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#3-矩阵构建及差异分析"><span class="nav-number">3.</span> <span class="nav-text">3. 矩阵构建及差异分析</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#3-1-表达矩阵"><span class="nav-number">3.1.</span> <span class="nav-text">3.1 表达矩阵</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#3-1-1-读取文件"><span class="nav-number">3.1.1.</span> <span class="nav-text">3.1.1 读取文件</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-1-2-接着构建DGEList对象"><span class="nav-number">3.1.2.</span> <span class="nav-text">3.1.2 接着构建DGEList对象</span></a></li></ol></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#4-数据过滤"><span class="nav-number">4.</span> <span class="nav-text">4. 数据过滤</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#5-标准化"><span class="nav-number">5.</span> <span class="nav-text">5. 标准化</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#6-差异表达分析"><span class="nav-number">6.</span> <span class="nav-text">6. 差异表达分析</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      
        <div class="back-to-top">
          <i class="fa fa-arrow-up"></i>
          
            <span id="scrollpercent"><span>0</span>%</span>
          
        </div>
      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2020</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">WENJU QIAN</span>

  
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-area-chart"></i>
    </span>
    
      <span class="post-meta-item-text">Site words total count&#58;</span>
    
    <span title="Site words total count">10k</span>
  
</div>


  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Gemini</a> v5.1.4</div>




# 添加访客数
<div class="powered-by">
	<i class="fa fa-user-md"></i>
	<span id="busuanzi_container_site_uv">
		本站访客数:<span id="busuanzi_value_site_uv"></span>
	</span>
	<span class="post-meta-divider">|</span>
	<span id="busuanzi_container_site_pv">
		本站访问量<span id="busuanzi_value_site_pv"></span>
	</span>
</div>

# 统计博客总计运行时间
<span id="sitetime"></span>
<script language=javascript>
  function siteTime(){
    window.setTimeout("siteTime()", 1000);
    var seconds = 1000;
    var minutes = seconds * 60;
    var hours = minutes * 60;
    var days = hours * 24;
    var years = days * 365;
    var today = new Date();
    var todayYear = today.getFullYear();
    var todayMonth = today.getMonth()+1;
    var todayDate = today.getDate();
    var todayHour = today.getHours();
    var todayMinute = today.getMinutes();
    var todaySecond = today.getSeconds();
    /* Date.UTC() -- 返回date对象距世界标准时间(UTC)1970年1月1日午夜之间的毫秒数(时间戳)
    year - 作为date对象的年份，为4位年份值
    month - 0-11之间的整数，做为date对象的月份
    day - 1-31之间的整数，做为date对象的天数
    hours - 0(午夜24点)-23之间的整数，做为date对象的小时数
    minutes - 0-59之间的整数，做为date对象的分钟数
    seconds - 0-59之间的整数，做为date对象的秒数
    microseconds - 0-999之间的整数，做为date对象的毫秒数 */
    var t1 = Date.UTC(2020,03,15,00,00,00); //这里调整博客建站时间，时间：2020-03-15 00:00:00
    var t2 = Date.UTC(todayYear,todayMonth,todayDate,todayHour,todayMinute,todaySecond);
    var diff = t2-t1;
    var diffYears = Math.floor(diff/years);
    var diffDays = Math.floor((diff/days)-diffYears*365);
    var diffHours = Math.floor((diff-(diffYears*365+diffDays)*days)/hours);
    var diffMinutes = Math.floor((diff-(diffYears*365+diffDays)*days-diffHours*hours)/minutes);
    var diffSeconds = Math.floor((diff-(diffYears*365+diffDays)*days-diffHours*hours-diffMinutes*minutes)/seconds);
    document.getElementById("sitetime").innerHTML=" 已运行"/*+diffYears+" 年 "*/+diffDays+" 天 "+diffHours+" 小时 "+diffMinutes+" 分钟 "+diffSeconds+" 秒";
  }/*因为建站时间还没有一年，就将之注释掉了。需要的可以取消*/
  siteTime();
</script>
        
<div class="busuanzi-count">
  <script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="site-uv">
      <i class="fa fa-user"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
      人次
    </span>
  

  
    <span class="site-pv">
      <i class="fa fa-eye"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
      次
    </span>
  
</div>








        
      </div>
    </footer>

    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  


  











  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>
  

  
  
    <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>
  

  
  
    <script type="text/javascript" src="/lib/canvas-nest/canvas-nest.min.js"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.4"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.4"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.4"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.4"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.4"></script>



  


  




	





  





  










  <script src="//cdn1.lncld.net/static/js/3.0.4/av-min.js"></script>
  <script src="//unpkg.com/valine/dist/Valine.min.js"></script>
  
  <script type="text/javascript">
    var GUEST = ['nick','mail','link'];
    var guest = 'nick,mail,link';
    guest = guest.split(',').filter(item=>{
      return GUEST.indexOf(item)>-1;
    });
    new Valine({
        el: '#comments' ,
        verify: true,
        notify: true,
        appId: '0uYeC1X5ME3X6vtooIJoYito-gzGzoHsz',
        appKey: '7u586HL29S9uVvurXaAzOhfn',
        placeholder: '有想说的欢迎留言交流，留下你的邮箱，我会尽快回复~（邮箱不会被公开）',
        avatar:'mm',
        guest_info:guest,
        pageSize:'10' || 10,
    });
  </script>



  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  
  

  

  

  

<script src="/live2dw/lib/L2Dwidget.min.js?094cbace49a39548bed64abff5988b05"></script><script>L2Dwidget.init({"model":{"jsonPath":"/live2dw/assets/hijiki.model.json"},"display":{"position":"right","width":140,"height":260},"mobile":{"show":true},"log":false,"pluginJsPath":"lib/","pluginModelPath":"assets/","pluginRootPath":"live2dw/","tagMode":false});</script></body>
</html>
